crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl

```diff
@@ -0,0 +1,168 @@
+            // common in motion-vector-based motion blur implementations. While some resources
+            // recommend using depth, I've found that sampling the velocity results in significantly
+            // better results. Unlike a depth heuristic, this is not scale dependent.
+            //
+            // The most distracting artifacts occur when a stationary foreground object is
+            // incorrectly sampled while blurring a moving background object, causing the stationary
+            // object to blur when it should be sharp ("background bleeding"). This is most obvious
+            // when the camera is tracking a fast moving object. The tracked object should be sharp,
+            // and should not bleed into the motion blurred background.
+            //
+            // To attenuate these incorrect samples, we compare the motion of the fragment being
+            // blurred to the UV being sampled, to answer the question "is it possible that this
+            // sample was occluding the fragment?"
+            //
+            // Note to future maintainers: proceed with caution when making any changes here, and
+            // ensure you check all oclusion/disocclusion scenarios and fullscreen camera rotation
+            // blur for regressions.
+            let frag_speed = length(step_vector);
+            let sample_speed = length(sample_motion) / 2.0; // Halved because the sample is centered
+            let cos_angle = dot(step_vector, sample_motion) / (frag_speed * sample_speed * 2.0);
+            let motion_similarity = clamp(abs(cos_angle), 0.0, 1.0);
+            if sample_speed * motion_similarity < frag_speed {
+                // Project the sample's motion onto the frag's motion vector. If the sample did not
+                // cover enough distance to reach the original frag, there is no way it could have
+                // influenced this frag at all, and should be discarded.
+                weight = 0.0;
+            }
+        }
+        weight_total += weight;
+        accumulator += weight * sample_color;
+    }
+
+    let has_moved_less_than_a_pixel = length(this_motion_vector * texture_size) < 1.0;
+    // In case no samples were accepted, fall back to base color.
+    // We also fall back if motion is small, to not break antialiasing.
+    if weight_total <= 0.0 || has_moved_less_than_a_pixel {
+        accumulator = base_color;
+        weight_total = 1.0;
+    }
+    return accumulator / weight_total;
+}
+
+fn hash_noise(ifrag_coord: vec2<i32>, frame: u32) -> f32 {
+    let urnd = uhash(u32(ifrag_coord.x), (u32(ifrag_coord.y) << 11u) + frame);
+    return unormf(urnd);
+}
+
+fn uhash(a: u32, b: u32) -> u32 { 
+    var x = ((a * 1597334673u) ^ (b * 3812015801u));
+    // from https://nullprogram.com/blog/2018/07/31/
+    x = x ^ (x >> 16u);
+    x = x * 0x7feb352du;
+    x = x ^ (x >> 15u);
+    x = x * 0x846ca68bu;
+    x = x ^ (x >> 16u);
+    return x;
+}
```

Comment:
I think these should be moved to `utils.wgsl`, as that's where our other rand-like functions live.